sycl/source/detail/xpti_registry.hpp

```diff
@@ -111,6 +112,122 @@ class XPTIRegistry {
                    uint16_t TraceEventType);
 #endif // XPTI_ENABLE_INSTRUMENTATION
 };
+
+/// @brief Helper class to enable XPTI implementation
+/// @details This class simplifies the instrumentation and encapsulates the
+/// verbose call sequences
+class XPTIScope {
+public:
+  using TracePoint = xpti::framework::tracepoint_t;
+  /// @brief Scoped class for XPTI instrumentation using TLS data
+  /// @param CodePtr  The address of the class/function to help differentiate
+  /// actions in case the code location information is not available
+  /// @param TraceType The type of trace event being created
+  /// @param StreamName  The stream which will emit these notifications
+  /// @param UserData String value that provides metadata about the
+  /// instrumentation
+  XPTIScope(void *CodePtr, uint16_t TraceType, const char *StreamName,
+            const char *UserData)
+      : MUserData(UserData), MStreamID(0), MInstanceID(0), MScopedNotify(false),
+        MTraceType(0) {
+#if XPTI_ENABLE_INSTRUMENTATION
+    detail::tls_code_loc_t Tls;
+    auto TData = Tls.query();
+    // If TLS is not set, we can still genertate universal IDs with user data
+    // and CodePtr information
+    const char *FuncName = UserData;
+    if (TData.functionName())
+      FuncName = TData.functionName();
+    // Create a tracepoint object that has a lifetime of this class
+    MTP = new TracePoint(TData.fileName(), FuncName, TData.lineNumber(),
+                         TData.columnNumber(), CodePtr);
+    if (MTraceType == (uint16_t)xpti::trace_point_type_t::graph_create ||
+        MTraceType == (uint16_t)xpti::trace_point_type_t::node_create ||
+        MTraceType == (uint16_t)xpti::trace_point_type_t::edge_create)
+      MTP->parent_event(GSYCLGraphEvent);
+    // Now if tracing is enabled, create trace events and notify
+    if (xptiTraceEnabled() && MTP) {
+      MTP->stream(StreamName).trace_type((xpti::trace_point_type_t)TraceType);
+      MTraceEvent = const_cast<xpti::trace_event_data_t *>(MTP->trace_event());
+      MStreamID = MTP->stream_id();
+      MInstanceID = MTP->instance_id();
+    }
+#endif
+  }
+
+  xpti::trace_event_data_t *traceEvent() { return MTraceEvent; }
+
+  uint8_t streamID() { return MStreamID; }
+
+  uint64_t instanceID() { return MTP ? MTP->instance_id() : 0; }
+
+  XPTIScope &
+  addMetadata(const std::function<void(xpti::trace_event_data_t *)> &Callback) {
```

Comment:
That answers my concern about `(auto _&_ TEvent)`, although I'm not sure if I like that answer. I'd still prefer to introduce a helper wrapper class for that `TEvent` param (see my other comment about `addMetadata` in such a wrapper).